File: Stacks_Queues/go/sort.go
package main

import (
	"types"
)


func insertTheSmallest(s *types.Stack, tempStack *types.Stack, temp int) {
  var top, topTemp, nums int
  for {
    if tempStack.IsEmpty() {
      break
    }

    topTemp, _ = tempStack.Peek()
    if temp >= topTemp {
      break
    }

    top, _ = tempStack.Pop()
    s.Push(top)
    nums++
  }

  tempStack.Push(temp)

  for i := 0; i < nums; i++ {
    top, _ = s.Pop()
    tempStack.Push(top)
  }
}


func Sort(s *types.Stack) {
	tempStack := &types.Stack{}
	var top, topTemp int

	for !s.IsEmpty() {
		top, _ = s.Pop()

		if tempStack.IsEmpty() {
			tempStack.Push(top)
		}

		topTemp, _ = tempStack.Peek()
		if top >= topTemp {
			tempStack.Push(top)
		} else {
			insertTheSmallest(s, tempStack, top)
		}
	}

	for !tempStack.IsEmpty() {
		top, _ = tempStack.Pop()
		s.Push(top)
	}

}

func main() {
	var s types.Stack
	s.Push(1)
	s.Push(-23)
	s.Push(0)
	s.Push(3)
	s.Push(2)
	s.Push(10)
	Sort(&s)
	types.PrintStack(s)
}

File: Stacks_Queues/python/sort.py
from Stack import Stack


class SortingStack(Stack):
    def sort(self):
        temp_stack = Stack()

        while not self.is_empty():
            top = self.pop()
            temp = top
            nums = 0

            if temp_stack.is_empty():
                temp_stack.push(top)
                continue

            top_temp = temp_stack.peek()

            if top >= top_temp:
                temp_stack.push(top)
            else:
                while True:
                    if temp_stack.is_empty():
                        break

                    top_temp = temp_stack.peek()

                    if temp >= top_temp:
                        break

                    top = temp_stack.pop()
                    self.push(top)
                    nums += 1

                temp_stack.push(temp)

                for i in range(nums):
                    top = self.pop()
                    temp_stack.push(top)

        while not temp_stack.is_empty():
            top = temp_stack.pop()
            self.push(top)


def main():
    s = SortingStack()
    s.push(1)
    s.push(-23)
    s.push(0)
    s.push(3)
    s.push(2)
    s.push(10)

    print(s)
    s.sort()
    print(s)

if __name__ == '__main__':
    main()
